### DISCUSSION TITLE
configuring objects with shared parameters

### DISCUSSION THREAD

**Discussion by dagap** [facebookresearch/hydra#2401] [Q&A]
I have multiple functions that have shared parameters. something like:

```
def x(param_1: int, param_2: int, param_3: int, **kwargs)
```

and another one like:
```
def y(param_1: int, param_2: int, param_3: int, **kwargs)
```

I was wondering how I can configure something where I can have these shared parameters:

Something like:

```
# shared_values.yaml
param_1: <value>
param_2: <value>
param_3: <value>
```

```
# x.yaml
x:
   other_params: <value>
   # how can I inline the values here from the other file
```
   
```
# y.yaml
y:
   other_params: <value>
   # how can I inline the values here from the other file
```

Is there a way to do it in this paradigm or does anyone have a suggestion how one can achieve something like this?

**Comment by omry** [comment]
Check the [Extending Configs](https://hydra.cc/docs/patterns/extending_configs/) pattern.

Note: this builds in some advanced usage of the defaults list. I strongly recommend that you spend the time reading and understanding [The Defaults List](https://hydra.cc/docs/advanced/defaults_list/).

Related:
If you model your config differently, you could also use OmegaConf interpolation as a simpler solution (which does not meet the requirements you described).

**Reply by dagap** [reply]
Hi @omry. Thanks for the reply. There is a whole set of issues with this approach. The problem is that the defaults list get overwritten when using multiple items of the list. Take this simple example:

```
# params.yaml

threshold: 100
```

```
# a.yaml

defaults:
  - params.yaml
```

```
# b.yaml

defaults:
  - params.yaml
```

```
# combined.yaml

defaults:
  - a
  - b
```

Now, the combined is going to merge the parameters. The issue is that the defaults list need to be at the top level and this is hurting when you want to share parameters across instances.

It would be nice to have something like:

```
# a.yaml

a:
  - defaults:
     - params
```
At least this way it would be preserved but I am not sure how to achieve the same thing.

**Reply by omry** [reply]
> Now, the combined is going to merge the parameters. The issue is that the defaults list need to be at the top level and this is hurting when you want to share parameters across instances.

The defaults list needs to be at the top level indeed, and this is not likely to ever change.
You can, however, control the package (node path) a config is composed into. Check the [packages](https://hydra.cc/docs/advanced/overriding_packages/) page.
